rust/lance/src/io/commit/s3_test.rs

```diff
@@ -0,0 +1,345 @@
+        };
+    }
+}
+
+impl Drop for DynamoDBCommitTable {
+    fn drop(&mut self) {
+        let table_name = self.0.clone();
+        tokio::task::spawn(async move {
+            let config = aws_config().await;
+            let client = aws_sdk_dynamodb::Client::new(&config);
+            Self::delete_table(client, &table_name).await;
+        });
+    }
+}
+
+#[tokio::test]
+async fn test_concurrent_writers() {
+    use crate::utils::test::IoTrackingStore;
+
+    let datagen = gen().col("values", array::step::<Int32Type>());
+    let data = datagen.into_batch_rows(RowCount::from(100)).unwrap();
+
+    let (io_stats_wrapper, io_stats) = IoTrackingStore::new_wrapper();
+
+    // Create a table
+    let store_params = ObjectStoreParams {
+        object_store_wrapper: Some(io_stats_wrapper),
+        storage_options: Some(
+            CONFIG
+                .iter()
+                .map(|(k, v)| (k.to_string(), v.to_string()))
+                .collect(),
+        ),
+        ..Default::default()
+    };
+    let write_params = WriteParams {
+        store_params: Some(store_params.clone()),
+        mode: WriteMode::Append,
+        ..Default::default()
+    };
+    let bucket = S3Bucket::new("test-concurrent-writers").await;
+    let uri = format!("s3://{}/test", bucket.0);
+    let transaction = InsertBuilder::new(&uri)
+        .with_params(&write_params)
+        .execute_uncommitted(vec![data.clone()])
+        .await
+        .unwrap();
+
+    // 1 IOPS for uncommitted write
+    let incremental_stats = || {
+        let mut stats = io_stats.as_ref().lock().unwrap();
+        std::mem::take(stats.deref_mut())
+    };
+    assert_eq!(incremental_stats().write_iops, 1);
+
+    let dataset = CommitBuilder::new(&uri)
+        .with_store_params(store_params.clone())
+        .execute(transaction)
+        .await
+        .unwrap();
+    // Commit: 2 IOPs. 1 for transaction file, 1 for manifest file
+    assert_eq!(incremental_stats().write_iops, 2);
+    let dataset = Arc::new(dataset);
+    let old_version = dataset.manifest().version;
+
+    let concurrency = 10;
+    let mut tasks = Vec::with_capacity(concurrency);
+    for _ in 0..concurrency {
+        let ds_ref = dataset.clone();
+        let data_ref = data.clone();
+        let task = tokio::spawn(async move {
+            InsertBuilder::new(ds_ref)
+                .with_params(&WriteParams {
+                    mode: WriteMode::Append,
+                    ..Default::default()
+                })
+                .execute(vec![data_ref])
+                .await
+                .unwrap();
+        });
+        tasks.push(task);
+    }
+    try_join_all(tasks).await.unwrap();
+
+    let mut dataset = dataset.as_ref().clone();
+    dataset.checkout_latest().await.unwrap();
+    assert_eq!(old_version + concurrency as u64, dataset.manifest().version);
+
+    let num_rows = dataset.count_rows(None).await.unwrap();
+    assert_eq!(num_rows, data.num_rows() * (concurrency + 1));
```

Comment:
Maybe do a quick scan to make sure the data is all readable?  Or perhaps even better, a call to dataset validate?